Close #13620: Add patrol area retrieval method for plugins

csunday95 opened the pull request:
- allows a script engine plugin to conveniently get the staff patrol
   area matrix for a particular staff member
 - returns a javascript Array of Arrays of bools representing a 64x64
   grid

duncanspumpkin commented on src/openrct2/scripting/ScMap.hpp:
```diff
@@ -168,6 +191,7 @@ namespace OpenRCT2::Scripting
             dukglue_register_method(ctx, &ScMap::getTile, "getTile");
             dukglue_register_method(ctx, &ScMap::getEntity, "getEntity");
             dukglue_register_method(ctx, &ScMap::getAllEntities, "getAllEntities");
+            dukglue_register_method(ctx, &ScMap::getStaffPatrolArea, "getStaffPatrolArea");
```
Why would you make this part of map and not part of entity?

csunday95 replied:
Adding getPatrolArea as a dukglue method on the SCStaff type does make more sense; making that modification now. I'm also not clear on which part of the commit message guide I'm not following; just the keep it in present tense part?

duncanspumpkin commented:
Please follow https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages for your commit messages.

IntelOrca commented:
I was hoping your function would return a list of tile co-coordinates that are in their patrol area. Returning a large boolean array is probably going to be a lot more inefficient. It might even be better to calculate rectangle ranges for the patrol area to help improve the performance.

It would probably also be useful to have a set of functions on a patrol area object. E.g.
```ts
interface PatrolArea {
    tiles: CoordsXY[];

    add(coords: CoordsXY[]): void;
    remove(coords: CoordsXY[]): void;
    contains(coord: CoordsXY): bool;
}

interface Staff extends Peep {
    // existing stuff ...

    readonly patrolArea: PatrolArea;
}
```

csunday95 commented:
I've moved the getPatrolArea method to the ScStaff object. I've also added a hasPatrolArea property to disambiguate the situation where the staff has "no patrol area", which is equivalent to having every possible location in its patrol area. I can also change to this to returning bounding rectangles as an Array of MapRange (with the topleft/bottom right map coordinates as the points), or create a different method for that.

csunday95 commented:
Also, as it currently stands, a plugin will need to use scripting engine actions to do tasks like setting staff patrol area. E.g.:

```js
context.executeAction("staffsetpatrolarea", {
    peep: peep.id,
    loc: locationData
}, function (result) {...handle result...});
```

Would it be preferred to implement more intermediary duk methods, e.g. create a setPatrolArea function on ScStaff that handles creating/executing the requisite action(s)

saeltz commented:
> Would it be preferred to implement more intermediary duk methods, e.g. create a setPatrolArea function on ScStaff that handles creating/executing the requisite action(s)?

From a plugin author point of view this is much preferred.

csunday95 commented:
okay, this latest iteration now returns an Array of TileRange objects, which specify the corners of the patrol region in tile coordinates. I've added a definition for it to the TypeScript file as

```ts
interface TileRange {
    x1: number;
    y1: number;
    x2: number; 
    y2: number;
}
```

Currently, I have the range as _inclusive_, i.e. a tile range for a patrol region will look like

```js
{x1: 4, y1: 8, x2: 7, y2: 11}
```

but switching it to the exclusive upper bound convention is trivial. I wasn't sure which convention is used for MapRange etc currently.

duncanspumpkin commented on src/openrct2/scripting/ScEntity.hpp:
```diff
@@ -1188,6 +1190,39 @@ namespace OpenRCT2::Scripting
+            }
+            for (int32_t y = 0; y < 64; y++)
+            {
+                auto y_index = y << 7;
```
don't use shifts unless you absolutely have to. Leave it to the compiler to optimise them.

duncanspumpkin replied:
Also you could have just done +128 instead of ++ to avoid the multiplication.

csunday95 replied:
I was simply copying the shift logic used in the Staff class for consistency, but there's myriad ways to write this code of course. The most "readable" implementation, I think, would be keeping the single ++'s and using expanded multiplication like 
`y_index = y * 32 * 4;`

Gymnasiast replied:
The 32 is for the number of big x/y coordinates in 1 small coordinate. We have a constant for that: `COORDS_XY_STEP`. 

The 4 is for the number of tiles in a single piece of patrol area. I don't know if we have a constant for it, but if we haven't, please introduce one.

duncanspumpkin commented on src/openrct2/world/Location.hpp:
```diff
@@ -729,6 +729,16 @@ struct MapRange : public RectRange<CoordsXY>
     }
 };
 
+struct TileRange : public RectRange<TileCoordsXY>
```
We are trying to remove more instances of TileCoordsXY not add more. Leave in terms of `CoordsXY` (i.e. just use MapRange)

csunday95 replied:
Well I see using MapRange in this scenario as going one of two ways:

1) plugin writers have to be aware of the 32x resolution per tile, i.e. when looking at a tile with the file inspector, the x/y values have to be multiplied by 32
2) OR we begin using MapRange with a completely different basis.

What's the rationale behind trying to remove TileCoordsXY? it seems logical to me to have both classes for the different coordinate systems

Gymnasiast replied:
Mostly to avoid endlessly converting between the two. Entities always use CoordsXY, tile elements use a mixture. (Don't get us started on the numerous places that use X and Y from TileCoordsXYZ, but Z from CoordsXYZ...)

csunday95 replied:
so for the plugin interface, should we still send out serialized coordinates in the CoordsXY space (i.e. 32 steps per tile) or in tile space? I think very few plugins would actually make use of the higher precision coordinate system, but perhaps some.

duncanspumpkin replied:
Everything at present plugin wise uses CoordsXY as far as I know. There shouldn't be anything that works in terms of TileCoordsXY.

IntelOrca commented:
> Would it be preferred to implement more intermediary duk methods, e.g. create a setPatrolArea function on ScStaff that handles creating/executing the requisite action(s)?

Currently only some of the network commands actually execute game actions. This is mainly for convenience and they are a special case of game actions.

The game actions are designed around how the user edits the patrol areas. Plugins should instead create their own custom action which edits the patrol area more efficiently using a more direct API. This allows you to send one game action to set a large patrol area which would usually take many game actions.

I think it is important to get the API right first time otherwise it causes a lot of issues for existing plugins when we need to add extra functionality requiring the API to change. The API I mocked up above provides a simpler method for checking individual tile while allowing a batch of tiles to be added or removed. It will still behave correctly when we finally allow patrol areas to be smaller than 4x4 regions. What are your thoughts? Are you able go further and implement some kind of ScPatrolArea class or would you prefer me to take over?

csunday95 commented:
@IntelOrca I'm willing to implement as elaborate as a system as needed, I was simply trying to create a minimal system that would work without making too many assumptions as to what was required/desired. Creating a new Duktape class as you suggested seems like a good way to go, other than it also possibly needing to retain the staff sprite ID (unless we want PatrolArea 's to be entirely "abstract" and not associated with any particular staff member). So for clarity, the general consensus so far seems to be:

1) Create a new ScPatrolArea class for duktape
   - this class has an array of MapRange objects as its member data, possibly merged to minimize the number of array entries
   - has `Add()`, `Remove()` and `Contains()` methods 
   - Question: I think for efficiency, we would want add/remove/contain to execute on the plugin side and commit the new data to a staff member only when local operations are complete? 
 
2) make the getPatrolArea method on the ScStaff class return an ScPatrolArea instance
   - Question: do we want ScPatrolArea's to be independent of a particular staff member, or possibly store a staff sprite ID they correspond to?
 
3) >Plugins should instead create their own custom action which edits the patrol area more efficiently using a more direct API. This allows you to send one game action to set a large patrol area which would usually take many game actions.

 - so if I'm understanding correctly in this case we would create a `SetPatrolArea` method to complement the `Get` one on the ScStaff class. This `SetPatrolArea` would itself take a ScPatrolArea instance as its argument, and itself call `staff_toggle_patrol_area()` as needed/perform the required checks?

IntelOrca commented:
I think `ScPatrolArea` should be created on-demand when `staffMember.patrolArea` is accessed. `ScPatrolArea` has a peep ID associated with it and it uses that alone to read/edit the patrol area. Given the add and remove members take one or more coordinates, you can commit after each call.

csunday95 commented:
>I think ScPatrolArea should be created on-demand when staffMember.patrolArea is accessed

So something like a lazy property getter method in the typescript definitions? It appears that the openrct2.d.ts file is entirely declarative now, would we want to start adding some actual executable code to it. Otherwise I'm not sure how you'd do a lazy evaluation type property. 
```

IntelOrca commented:
@csunday95 No I just mean the getter for it in the C++ code would return a new instance of ScPatrolArea. Have a look at how map elements work or the scenario options.

csunday95 commented on src/openrct2/scripting/ScEntity.hpp:
```diff
@@ -1057,6 +1057,251 @@ namespace OpenRCT2::Scripting
+            // produce minimum possible number of rectangular regions (though not
+            // necessarily the most "square" ones)
+            constexpr int32_t PATROL_REGION_STEP = COORDS_XY_STEP * COORDS_XY_PATROL_STEP;
+            std::array<std::array<MapRange*, 64>, 64> patrolBitmask = { nullptr };
```
quite a large stack allocation; could be replaced with a `std::vector<std::vector<MapRange*>>` and populated in place on line 1138 below

csunday95 commented on src/openrct2/scripting/ScEntity.hpp:
```diff
@@ -1057,6 +1057,251 @@ namespace OpenRCT2::Scripting
+            }
+        
+            // horizontal scanning
+            for (int32_t y = 0; y < 64; y++)
```
nested loops can be collapsed to a big loop with calculated indexing

tupaschoal commented:
Is this just missing the compilation error fixes or is it still a WIP?

csunday95 commented:
This is still WIP; I need to implement the add/remove functions and create some unit tests for the merging algorithm. I was pretty busy during the holiday season, but I should have some time to work on it today and this week

IntelOrca commented:
I've labelled this as NSF, as that will change patrol areas to single tiles. It should wait until then and make use of the utility methods that will come with it.
